src/multi_van.h

```diff
@@ -0,0 +1,273 @@
+    for (auto& thread : polling_threads_) {
+      thread->join();
+      thread.reset();
+    }
+
+    // stop zmq vans
+    for (auto van : vans_) {
+      van->Stop();
+    }
+    PS_VLOG(1) << "polling_threads_ stopped";
+  }
+
+  int Bind(Node& node, int max_retry) override {
+    std::lock_guard<std::mutex> lk(endpoints_mu_);
+    CHECK_EQ(node.num_ports, num_ports_);
+    my_nodes_.resize(num_ports_);
+    for (int i = 0; i < num_ports_; ++i) {
+      int port = node.ports[i];
+      Node one_node = node;
+      one_node.id = EncodeManagedID(node.id, i);
+      one_node.num_ports = 1;
+      one_node.ports[0] = port;
+      one_node.port = node.port;
+      one_node.dev_types[i] = CPU;
+      one_node.dev_ids[i] = i;
+      int bound_port = vans_[i]->Bind(one_node, max_retry);
+
+      // update node with port and device info
+      node.dev_types[i] = CPU;
+      node.dev_ids[i] = i;
+      node.ports[i] = bound_port;
+      my_nodes_[i] = one_node;
+      polling_threads_.emplace_back(new std::thread(&MultiVan::PollingThread, this, i));
+    }
+    return node.ports[0];
+  }
+
+  void Connect(const Node &node) override {
+    CHECK_NE(node.id, node.kEmpty);
+    CHECK_NE(node.port, node.kEmpty);
+    CHECK(node.hostname.size());
+
+    // worker doesn't need to connect to the other workers. same for server
+    if ((node.role == my_node_.role) && (node.id != my_node_.id)) {
+      PS_VLOG(3) << "Multivan skipped connection to " << node.DebugString()
+                 << ". My node is " << my_node_.DebugString();
+      return;
+    }
+    for (int i = 0; i < num_ports_; ++i) {
+      for (int j = 0; j < node.num_ports; ++j) {
+        // connect each van pair
+        Node one_node = node;
+        // the underlying van uses id: 0001, 1001, 2001, etc
+        one_node.id = EncodeManagedID(node.id, j);
+        one_node.num_ports = 1;
+        one_node.port = node.ports[j];
+        one_node.ports[0] = one_node.port;
+        PS_VLOG(3) << "Connect: " << one_node.DebugString() << " from " << my_nodes_[i].DebugString();
+        vans_[i]->Connect(one_node);
+      }
+    }
+  }
+
+  int SendMsg(Message &msg) override {
+    int remote_id = msg.meta.recver;
+    CHECK_NE(remote_id, Node::kEmpty);
+    // XXX assume device IDs are: [0, 1 ... num_ports - 1]
+    bool pushpull = IsValidPushpull(msg);
+    int src_idx = 0;
+    int dst_idx = 0;
+    if (pushpull && msg.data.size() == 3) {
+      auto& data = msg.data[1];
+      CHECK(data.src_device_type_ == CPU);
+      CHECK(data.dst_device_type_ == CPU);
+      src_idx = data.src_device_id_;
+      dst_idx = data.dst_device_id_;
+      CHECK_EQ(my_nodes_[src_idx].dev_types[0], data.src_device_type_);
+      CHECK_EQ(my_nodes_[src_idx].dev_ids[0], data.src_device_id_);
+    }
+    Message van_msg = msg;
+    auto van = vans_[src_idx];
+    van_msg.meta.sender = EncodeManagedID(msg.meta.sender, src_idx);
+    van_msg.meta.recver = EncodeManagedID(msg.meta.recver, dst_idx);
```

Comment:
BTW, should messages sent from different GPUs memory be treated as sent from different workers?

Reply:
Currently they are all treated as the same worker, and we rely on the device id/type in msg.meta to identify which device this data is from. Did you find the device info in msg.meta not sufficient for any case?

Reply:
I meant the scenario when the same worker sends different data from different GPUs, but with the same key. Is this situation possible?